goldmane/pkg/aggregator/index.go

```diff
@@ -60,42 +64,81 @@ type IndexFindOpts struct {
-	// We need to iterate through the DiachronicFlows to find the starting point, since not every DiachronicFlow is necessarily a match
-	// for the given options.
-	var i int
-	if opts.page > 0 {
-		target := opts.page * opts.limit
-		for j := 0; j < len(idx.diachronics); j++ {
-			if idx.evaluate(idx.diachronics[j], opts) != nil {
-				i++
-			}
-			if int64(i) == target {
-				break
+	target := int(opts.page * opts.pageSize)
+
+	// Iterate through the DiachronicFlows and evaluate each one until we reach the limit or the end of the list.
+	for _, diachronic := range idx.diachronics {
+		flow := idx.evaluate(diachronic, opts)
+		if flow != nil {
+			// increment the count regardless of whether we're including the key, as we need a total matching count.
+			totalMatchedCount++
+
+			// The target represents the index of the last key on the previous page, so we need to ensure that we're above
+			// that index before we start collecting flows.
+			if totalMatchedCount > target {
+				// Only append the key if
+				if opts.pageSize == 0 || int64(len(matchedFlows)) < opts.pageSize {
+					matchedFlows = append(matchedFlows, flow)
+				}
 			}
 		}
 	}
 
+	return matchedFlows, calculatePageCount(totalMatchedCount, int(opts.pageSize))
+}
+
+// Keys retrieves a unique set of keys matching the given index options.
+func (idx *index[E]) Keys(opts IndexFindOpts) ([]E, int) {
+	logrus.WithFields(logrus.Fields{
+		"opts": opts,
+	}).Debug("Listing flows from index")
+
+	var matchedKeys []E
+	var totalMatchedCount int
+
+	target := int(opts.page * opts.pageSize)
+	var previousKey *E
+
 	// Iterate through the DiachronicFlows and evaluate each one until we reach the limit or the end of the list.
-	for ; i < len(idx.diachronics); i++ {
-		flow := idx.evaluate(idx.diachronics[i], opts)
+	for _, diachronic := range idx.diachronics {
+		flow := idx.evaluate(diachronic, opts)
 		if flow != nil {
-			matchedFlows = append(matchedFlows, flow)
-		}
-
-		if opts.limit > 0 && int64(len(matchedFlows)) == opts.limit {
-			return matchedFlows
+			key := idx.keyFunc(&diachronic.Key)
+			// If the previous key does not equal the current key we know that we haven't seen this key yet, as this
+			// is sorted list and all keys with the same value are together.
+			if previousKey != nil && key == *previousKey {
+				// If we've seen this key match before then continue since it won't count as part of the count.
+				continue
+			}
+			previousKey = &key
+
+			// increment the count regardless of whether we're including the key, as we need a total matching count.
+			totalMatchedCount++
+
+			// The target represents the index of the last key on the previous page, so we need to ensure that we're above
+			// that index before we start collecting keys.
+			if totalMatchedCount > target {
+				// Only append the key if
+				if opts.pageSize == 0 || int64(len(matchedKeys)) < opts.pageSize {
```

Comment:
```suggestion
                        // Include the value if:
                        // - We're not performing a paginated search
                        // - We are performing a paginated search, and it falls within the page bounds.
			if opts.pageSize == 0 || totalMatchedCount > target || int64(len(matchedKeys)) < opts.pageSize {
```

I find this phrasing of the condition a bit easier to follow, but just preference.

Reply:
That's not correct though, it would should be `totalMatchedCount > pageStart && (opts.pageSize == 0 || int64(len(matchedFlows)) < opts.pageSize)` (which I can change it to).